include/swift/AST/Import.h

```diff
@@ -337,16 +399,296 @@ class ImportPath : public detail::ImportPathBase<ImportPath> {
   /// including submodules, assuming the \c ImportDecl has the indicated
   /// \c importKind.
   Module getModulePath(ImportKind importKind) const {
-    return getModulePath(importKind != ImportKind::Module);
+    return getModulePath(isScopedImportKind(importKind));
   }
 
   /// Extracts the portion of the \c ImportPath which represents a scope for the
   /// import, assuming the \c ImportDecl has the indicated \c importKind.
   Access getAccessPath(ImportKind importKind) const {
-    return getAccessPath(importKind != ImportKind::Module);
+    return getAccessPath(isScopedImportKind(importKind));
+  }
+};
+
+// MARK: - Abstractions of imports
+
+/// Convenience struct to keep track of an import path and whether or not it
+/// is scoped.
+class UnloadedImportedModule {
+  // This is basically an ArrayRef with a bit stolen from the size.
+  ImportPath::Raw::iterator data;
+  ImportPath::Raw::size_type isScoped : 1;
+  ImportPath::Raw::size_type length : sizeof(ImportPath::Raw::size_type)*CHAR_BIT - 1;
```

Comment:
IMO it would be better to stash the bit on the pointer – that way we can guarantee round-tripping the `ArrayRef`. What do you think?

Reply:
What happened here is that I felt that, if I did go to the trouble of stealing the extra bit from the pointer, I'd then want to extract the resulting logic into some kind of `ArrayRefIntPair` type, which would be an enormous pain. So instead I went in the other direction and stole the bit from the count, which was the easiest option.

This is probably a really stupid reason not to just implement what I need now and, y'know, *stop at that point*.

Reply:
Wouldn't it be sufficient to use a `PointerIntPair` for the `data` field?

Reply:
Yup, that's what I'll be doing.